File: python/drift/minifloat.py
"""An exact reference for the narrow float formats, written in rational arithmetic.

This is the other half of the oracle. The Java implementation in ``io.drift.fmt``
rounds by rescaling a double by a power of two, which is fast and exact but shares
its reasoning with everything downstream of it. This module reaches the same answers
by a different route: values are ``fractions.Fraction``, so a product of two
representable numbers is the true product with no rounding at all, and the only
approximation in the whole file is the single deliberate rounding under test.

Nothing here imports from the Java side or reads its output format definitions. The
field widths come from IEEE 754 and the OCP Microscaling specification. If the two
implementations agree on every code and every pair of codes, the agreement means
something.
"""

from __future__ import annotations

from dataclasses import dataclass
from fractions import Fraction
from typing import Union

# Sentinels for the values a Fraction cannot carry. Negative zero is one of them:
# every format here spends a code on it, and it is produced by ordinary arithmetic
# (a positive zero times any negative number), so leaving it out would make the
# reference disagree with the hardware on 154 of the 65,536 e4m3 products.
NAN = "nan"
POS_INF = "inf"
NEG_INF = "-inf"
NEG_ZERO = "-0"

Value = Union[Fraction, str]

IEEE = "IEEE"
FN = "FN"
FINITE = "FINITE"

SATURATE = "SATURATE"
PROPAGATE = "PROPAGATE"


def pow2(k: int) -> Fraction:
    """Exactly 2**k, for any sign of k."""
    return Fraction(2**k, 1) if k >= 0 else Fraction(1, 2**-k)


def floor_log2(a: Fraction) -> int:
    """The largest e with 2**e <= a, for a > 0. Pure integer reasoning."""
    if a <= 0:
        raise ValueError("floor_log2 needs a positive value")
    # bit_length gives the right answer up to one, so start there and walk.
    e = a.numerator.bit_length() - a.denominator.bit_length()
    while pow2(e) > a:
        e -= 1
    while pow2(e + 1) <= a:
        e += 1
    return e


def round_half_even(x: Fraction) -> int:
    """Nearest integer to an exact rational, ties to even."""
    n, d = x.numerator, x.denominator
    q, r = divmod(n, d)  # Python floors, which is what we want for negatives too
    twice = 2 * r
    if twice > d:
        q += 1
    elif twice == d and (q & 1):
        q += 1
    return q


@dataclass(frozen=True)
class Fmt:
    """A minifloat, described only by its fields and its treatment of the top exponent."""

    name: str
    exp_bits: int
    mant_bits: int
    bias: int
    specials: str
    overflow: str = SATURATE

    @property
    def width(self) -> int:
        return 1 + self.exp_bits + self.mant_bits

    @property
    def codes(self) -> int:
        return 1 << self.width

    @property
    def max_exp_field(self) -> int:
        return (1 << self.exp_bits) - 1

    @property
    def mant_mask(self) -> int:
        return (1 << self.mant_bits) - 1

    @property
    def sign_bit(self) -> int:
        return 1 << (self.exp_bits + self.mant_bits)

    @property
    def min_normal_exp(self) -> int:
        return 1 - self.bias

    @property
    def max_finite(self) -> Fraction:
        if self.specials == IEEE:
            top_exp, top_mant = self.max_exp_field - 1, self.mant_mask
        elif self.specials == FN:
            top_exp, top_mant = self.max_exp_field, self.mant_mask - 1
        else:
            top_exp, top_mant = self.max_exp_field, self.mant_mask
        sig = 1 + Fraction(top_mant, 1 << self.mant_bits)
        return sig * pow2(top_exp - self.bias)

    # ------------------------------------------------------------------ decode

    def decode(self, code: int) -> Value:
        """The exact value of a bit pattern."""
        if code < 0 or code >= self.codes:
            raise ValueError(f"code {code} outside {self.name}")
        mant = code & self.mant_mask
        exp_field = (code >> self.mant_bits) & self.max_exp_field
        negative = bool(code & self.sign_bit)

        if exp_field == self.max_exp_field:
            if self.specials == IEEE:
                if mant == 0:
                    return NEG_INF if negative else POS_INF
                return NAN
            if self.specials == FN and mant == self.mant_mask:
                return NAN

        if exp_field == 0:
            if mant == 0:
                return NEG_ZERO if negative else Fraction(0)
            value = Fraction(mant, 1) * pow2(self.min_normal_exp - self.mant_bits)
        else:
            sig = 1 + Fraction(mant, 1 << self.mant_bits)
            value = sig * pow2(exp_field - self.bias)
        return -value if negative else value

    # ------------------------------------------------------------------ encode

    def encode(self, value: Value) -> int:
        """Round an exact value into this format, nearest with ties to even."""
        if value is NAN or value == NAN:
            return self._nan_code()
        if value == POS_INF:
            return self._overflow_code()
        if value == NEG_INF:
            return self.sign_bit | self._overflow_code()
        if value == NEG_ZERO:
            return self.sign_bit

        assert isinstance(value, Fraction)
        sign = self.sign_bit if value < 0 else 0
        a = -value if value < 0 else value

        if a == 0:
            return sign

        e = floor_log2(a)
        if e < self.min_normal_exp:
            e = self.min_normal_exp

        quantum = pow2(e - self.mant_bits)
        m = round_half_even(a / quantum)

        normal_base = 1 << self.mant_bits
        if m < normal_base:
            # Subnormal; the integer is already the encoding.
            return sign | m
        if m >= normal_base * 2:
            m //= 2
            e += 1

        exp_field = e + self.bias
        mant_field = m - normal_base

        if exp_field > self.max_exp_field or (
            exp_field == self.max_exp_field and not self._top_binade_allows(mant_field)
        ):
            return sign | self._overflow_code()
        return sign | (exp_field << self.mant_bits) | mant_field

    def _top_binade_allows(self, mant_field: int) -> bool:
        if self.specials == IEEE:
            return False
        if self.specials == FN:
            return mant_field < self.mant_mask
        return True

    def _nan_code(self) -> int:
        if self.specials == IEEE:
            return (self.max_exp_field << self.mant_bits) | 1
        if self.specials == FN:
            return (self.max_exp_field << self.mant_bits) | self.mant_mask
        raise ArithmeticError(f"{self.name} has no NaN")

    def _overflow_code(self) -> int:
        if self.overflow == SATURATE:
            return self._max_finite_code()
        if self.specials == IEEE:
            return self.max_exp_field << self.mant_bits
        if self.specials == FN:
            return (self.max_exp_field << self.mant_bits) | self.mant_mask
        return self._max_finite_code()

    def _max_finite_code(self) -> int:
        if self.specials == IEEE:
            return ((self.max_exp_field - 1) << self.mant_bits) | self.mant_mask
        if self.specials == FN:
            return (self.max_exp_field << self.mant_bits) | (self.mant_mask - 1)
        return (self.max_exp_field << self.mant_bits) | self.mant_mask


E4M3 = Fmt("e4m3", 4, 3, 7, FN)
E5M2 = Fmt("e5m2", 5, 2, 15, IEEE)
E2M1 = Fmt("e2m1", 2, 1, 1, FINITE)
E3M2 = Fmt("e3m2", 3, 2, 3, FINITE)
E2M3 = Fmt("e2m3", 2, 3, 1, FINITE)
FP16 = Fmt("fp16", 5, 10, 15, IEEE)
BF16 = Fmt("bf16", 8, 7, 127, IEEE)

ALL = [E4M3, E5M2, E2M1, E3M2, E2M3, FP16, BF16]
BY_NAME = {f.name: f for f in ALL}


def is_nan(v: Value) -> bool:
    return isinstance(v, str) and v == NAN


def is_inf(v: Value) -> bool:
    return isinstance(v, str) and v in (POS_INF, NEG_INF)


def is_zero(v: Value) -> bool:
    return v == NEG_ZERO or (isinstance(v, Fraction) and v == 0)


def signbit(v: Value) -> bool:
    """The sign bit, which for zero is information a magnitude does not carry."""
    if isinstance(v, str):
        return v in (NEG_INF, NEG_ZERO)
    return v < 0


def magnitude(v: Value) -> Fraction:
    """The finite value as a rational, with negative zero flattened to zero."""
    if v == NEG_ZERO:
        return Fraction(0)
    assert isinstance(v, Fraction)
    return v


def add(fmt: Fmt, a: Value, b: Value) -> Value:
    """Exact sum, following IEEE 754 on the cases a rational cannot express."""
    if is_nan(a) or is_nan(b):
        return NAN
    if is_inf(a) or is_inf(b):
        if is_inf(a) and is_inf(b):
            return NAN if a != b else a
        return a if is_inf(a) else b
    if is_zero(a) and is_zero(b):
        # Two zeros sum to negative zero only when both were negative. Under any
        # other rounding mode than nearest this would not be true.
        return NEG_ZERO if (signbit(a) and signbit(b)) else Fraction(0)
    total = magnitude(a) + magnitude(b)
    # An exact cancellation is positive zero in round-to-nearest, whatever the signs.
    return total


def mul(fmt: Fmt, a: Value, b: Value) -> Value:
    """Exact product, following IEEE 754 on the cases a rational cannot express."""
    if is_nan(a) or is_nan(b):
        return NAN
    negative = signbit(a) ^ signbit(b)
    if is_inf(a) or is_inf(b):
        if is_zero(a) or is_zero(b):
            return NAN
        return NEG_INF if negative else POS_INF
    if is_zero(a) or is_zero(b):
        return NEG_ZERO if negative else Fraction(0)
    return magnitude(a) * magnitude(b)

File: java/src/main/java/io/drift/fmt/MiniFloat.java
package io.drift.fmt;

/**
 * A binary floating point format of arbitrary width, defined by its exponent and
 * mantissa field sizes and by what it does with the top exponent.
 *
 * <p>Every format used in low precision training is an instance of this: FP16 and
 * BF16 are IEEE-shaped, OCP E4M3 gives up infinity to buy one more binade, and FP4
 * E2M1 has no special values at all because with sixteen codes it cannot spare one.
 * Writing them once, parameterised, means the rounding path is a single piece of
 * code that gets exhaustively tested rather than five near-copies that do not.
 *
 * <p>The rounding is exact rather than approximate. The quantum at a given exponent
 * is a power of two, so dividing by it only shifts the exponent field of a double
 * and loses nothing; the scaled value is then an integer plus a fraction that a
 * double represents exactly. That is what lets {@code io.drift.tools.DumpTables} and
 * {@code python/verify_oracle.py} agree to the bit over every code in the format.
 */
public final class MiniFloat {

    /** What the all-ones exponent means. */
    public enum Specials {
        /** IEEE 754: top exponent is Inf (mantissa 0) or NaN (mantissa non-zero). */
        IEEE,
        /**
         * OCP "FN": finite, with NaN. Only the all-ones exponent with all-ones
         * mantissa is NaN. There is no infinity, which buys back every other code in
         * the top binade. That is how E4M3 reaches 448 instead of 240.
         */
        FN,
        /** No infinity and no NaN. Every bit pattern is a finite number. FP4 E2M1. */
        FINITE
    }

    /** What happens to a value larger than the format can hold. */
    public enum Overflow {
        /**
         * Clamp to the largest finite value. What every FP8 training recipe does,
         * because the alternative poisons the whole tensor.
         */
        SATURATE,
        /** Produce Inf or NaN as the format allows. IEEE behaviour. */
        PROPAGATE
    }

    public final String name;
    public final int expBits;
    public final int mantBits;
    public final int bias;
    public final Specials specials;
    public final Overflow overflow;

    /** Total width in bits, including the sign. */
    public final int width;
    private final int mantMask;
    private final int expMask;
    private final int signBit;
    private final int maxExpField;
    private final int codeMask;

    /** Unbiased exponent of the smallest normal value. */
    public final int minNormalExp;
    /** Largest finite magnitude. */
    public final double maxFinite;
    /** Smallest positive normal magnitude. */
    public final double minNormal;
    /** Smallest positive subnormal magnitude, the quantum in the subnormal range. */
    public final double minSubnormal;

    public MiniFloat(String name, int expBits, int mantBits, int bias,
                     Specials specials, Overflow overflow) {
        if (expBits < 1 || mantBits < 0 || expBits + mantBits > 30) {
            throw new IllegalArgumentException("unsupported field widths");
        }
        this.name = name;
        this.expBits = expBits;
        this.mantBits = mantBits;
        this.bias = bias;
        this.specials = specials;
        this.overflow = overflow;

        this.width = 1 + expBits + mantBits;
        this.mantMask = (1 << mantBits) - 1;
        this.maxExpField = (1 << expBits) - 1;
        this.expMask = maxExpField << mantBits;
        this.signBit = 1 << (expBits + mantBits);
        this.codeMask = (1 << width) - 1;

        this.minNormalExp = 1 - bias;
        this.minNormal = Math.scalb(1.0, minNormalExp);
        this.minSubnormal = Math.scalb(1.0, minNormalExp - mantBits);

        // The largest finite code depends on how much of the top binade the format
        // gave away to special values.
        int topExpField;
        int topMant;
        switch (specials) {
            case IEEE -> { topExpField = maxExpField - 1; topMant = mantMask; }
            case FN -> { topExpField = maxExpField; topMant = mantMask - 1; }
            case FINITE -> { topExpField = maxExpField; topMant = mantMask; }
            default -> throw new IllegalStateException();
        }
        if (topMant < 0) {
            throw new IllegalArgumentException("format too narrow for its special values");
        }
        this.maxFinite = Math.scalb(1.0 + (double) topMant / (1 << mantBits), topExpField - bias);
    }

    /** Number of distinct bit patterns, both signed zeros included. */
    public int codeCount() {
        return 1 << width;
    }

    // ---------------------------------------------------------------- decode

    /** The real value of a bit pattern. Patterns wider than the format are rejected. */
    public double decode(int bits) {
        if ((bits & ~codeMask) != 0) {
            throw new IllegalArgumentException("bits outside format width: " + bits);
        }
        int expField = (bits & expMask) >>> mantBits;
        int mant = bits & mantMask;
        double sign = (bits & signBit) != 0 ? -1.0 : 1.0;

        if (expField == maxExpField) {
            switch (specials) {
                case IEEE -> {
                    return mant == 0 ? sign * Double.POSITIVE_INFINITY : Double.NaN;
                }
                case FN -> {
                    if (mant == mantMask) return Double.NaN;
                }
                case FINITE -> { /* an ordinary number */ }
            }
        }
        if (expField == 0) {
            // Subnormal: no implicit leading one, exponent pinned at the minimum.
            return sign * Math.scalb((double) mant, minNormalExp - mantBits);
        }
        double significand = 1.0 + (double) mant / (1 << mantBits);
        return sign * Math.scalb(significand, expField - bias);
    }

    // ---------------------------------------------------------------- encode

    /** Encode with round-to-nearest-even. */
    public int encode(double v) {
        return encode(v, Rounding.NEAREST_EVEN, null);
    }

    /**
     * Encode a real value into this format.
     *
     * @param rng required for {@link Rounding#STOCHASTIC}, unused otherwise
     */
    public int encode(double v, Rounding rm, Rng rng) {
        if (Double.isNaN(v)) {
            if (specials == Specials.FINITE) {
                throw new ArithmeticException(name + " cannot represent NaN");
            }
            return nanBits();
        }

        int sign = (v < 0.0 || (v == 0.0 && 1.0 / v < 0.0)) ? signBit : 0;
        double a = Math.abs(v);

        if (a == 0.0) {
            return sign;
        }
        if (Double.isInfinite(a)) {
            return sign | overflowBits();
        }

        // Pick the binade, then the quantum inside it. Below the smallest normal the
        // quantum stops shrinking, which is the whole point of subnormals.
        int e = Math.getExponent(a);
        if (e == Double.MIN_EXPONENT - 1 || e < minNormalExp) {
            // Either a subnormal double, where the exponent query does not mean what
            // it usually does, or a value below this format's smallest normal. Both
            // are encoded against the fixed subnormal quantum.
            e = minNormalExp;
        }

        int quantumExp = e - mantBits;
        double scaled = Math.scalb(a, -quantumExp);   // exact: a power-of-two rescale
        long m = roundToIntegral(scaled, rm, rng);

        long normalBase = 1L << mantBits;

        if (m < normalBase) {
            // Subnormal. The raw integer is the encoding: if rounding pushed it up to
            // exactly normalBase it carries into the exponent field by itself, which
            // is the usual IEEE encoding trick and is handled by the branch below.
            return sign | (int) m;
        }
        if (m >= (normalBase << 1)) {
            // Rounding carried out of the binade. It can only land exactly on the
            // boundary, so halving is exact.
            m >>= 1;
            e += 1;
        }

        int expField = e + bias;
        int mantField = (int) (m - normalBase);

        if (expField > maxExpField
                || (expField == maxExpField && !topBinadeAllows(mantField))) {
            return sign | overflowBits();
        }
        return sign | (expField << mantBits) | mantField;
    }

    /** Round trip: the value this format would actually store for {@code v}. */
    public double quantize(double v) {
        return decode(encode(v, Rounding.NEAREST_EVEN, null));
    }

    /** Round trip under an explicit rounding mode. */
    public double quantize(double v, Rounding rm, Rng rng) {
        return decode(encode(v, rm, rng));
    }

    private boolean topBinadeAllows(int mantField) {
        return switch (specials) {
            case IEEE -> false;                  // top exponent is Inf/NaN only
            case FN -> mantField < mantMask;     // all-ones mantissa is NaN
            case FINITE -> true;
        };
    }

    private int nanBits() {
        return switch (specials) {
            case IEEE -> expMask | 1;
            case FN -> expMask | mantMask;
            case FINITE -> throw new ArithmeticException(name + " has no NaN");
        };
    }

    private int overflowBits() {
        if (overflow == Overflow.SATURATE) {
            return maxFiniteBits();
        }
        return switch (specials) {
            case IEEE -> expMask;                // +Inf
            case FN -> expMask | mantMask;       // no Inf exists, so NaN
            case FINITE -> maxFiniteBits();      // nothing to propagate to
        };
    }

    private int maxFiniteBits() {
        return switch (specials) {
            case IEEE -> ((maxExpField - 1) << mantBits) | mantMask;
            case FN -> (maxExpField << mantBits) | (mantMask - 1);
            case FINITE -> (maxExpField << mantBits) | mantMask;
        };
    }

    private static long roundToIntegral(double scaled, Rounding rm, Rng rng) {
        switch (rm) {
            case NEAREST_EVEN:
                return (long) Math.rint(scaled);
            case TRUNCATE:
                return (long) Math.floor(scaled);
            case STOCHASTIC: {
                double fl = Math.floor(scaled);
                double frac = scaled - fl;
                if (frac == 0.0) {
                    return (long) fl;
                }
                if (rng == null) {
                    throw new IllegalArgumentException("stochastic rounding needs an Rng");
                }
                return (long) fl + (rng.nextDouble() < frac ? 1 : 0);
            }
            default:
                throw new IllegalStateException();
        }
    }

    @Override
    public String toString() {
        return name;
    }
}
